File: games/spotted/sim/engine.py
"""SPOTTED! simulation: full rules engine.

Implements the rulebook in games/spotted/RULES.md exactly:
- draw 1, then one action (question / special / identification / discard)
- public answers increment the target's q counter (efficiency penalty meter)
- scoring: 2 + max(0, 4 - q) for a correct identification
- win at WIN_TARGET tokens; creature-deck exhaustion ends the round; tiebreaks
"""
from __future__ import annotations

import random
from dataclasses import dataclass, field

from .data import (
    Creature,
    QuestionCard,
    TRAITS,
    load_creatures,
    load_question_deck,
)

HAND_LIMIT = 7
STARTING_HAND = 6
WIN_TARGET_DEFAULT = 10
WIN_TARGET_2P = 8  # Duel variant
MAX_TURNS = 400  # safety cap


@dataclass
class PlayerState:
    seat: int
    hand: list[QuestionCard] = field(default_factory=list)
    specimen: Creature | None = None
    epoch: int = 0  # increments every time this player's specimen is replaced
    score: int = 0
    q: int = 0  # public answers about current specimen
    correct_ids: int = 0


@dataclass
class GameResult:
    winner: int  # seat, -1 for capped/abnormal
    turns: int
    scores: list[int]
    guess_attempts: int
    guess_hits: int
    q_at_hit: list[int] = field(default_factory=list)
    capped: bool = False


class Game:
    def __init__(self, n_players: int, agents, seed: int, win_target: int | None = None,
                 compensation: list[int] | None = None, rotate_first_player: bool = True):
        assert 2 <= n_players <= 5
        self.rng = random.Random(seed)
        self.n = n_players
        self.agents = agents
        self.rotate = rotate_first_player
        self.win_target = win_target or (WIN_TARGET_2P if n_players == 2 else WIN_TARGET_DEFAULT)
        self.creatures = load_creatures()
        self.creature_deck = self.creatures[:]
        self.rng.shuffle(self.creature_deck)
        self.question_deck = load_question_deck()
        self.rng.shuffle(self.question_deck)
        self.question_discard: list[QuestionCard] = []

        self.players = [PlayerState(seat=i) for i in range(n_players)]
        for p in self.players:
            p.specimen = self.creature_deck.pop()
            for _ in range(STARTING_HAND):
                p.hand.append(self.question_deck.pop())
            if compensation:
                # ACT-14 style advantage compensation: later seats act later in the
                # information race, so they start with tokens to balance win chances.
                # Values are calibrated by sim/calibrate.py.
                p.score = compensation[p.seat]

        # Fact log entries:
        #   ("answer", target, epoch, trait, value, yes, private_to)
        #   ("reveal", target, epoch, trait, value, private_to)
        #   ("guess_fail", target, epoch, cid)
        #   ("identified", target, epoch, cid, q_at_guess)
        self.log: list[tuple] = []
        self.depleted = False
        self.finish_triggered = False  # someone reached WIN_TARGET; finish the round
        self.turns = 0
        self.order = list(range(n_players))  # seat order for the current round
        self.pos = 0  # position within the round
        self.turn_seat = 0
        self.guess_attempts = 0
        self.guess_hits = 0
        self.q_at_hit: list[int] = []
        self.over = False
        self.winner = -1
        self.capped = False

    # ------------------------------------------------------------------ facts

    def public_q(self, seat: int) -> int:
        return self.players[seat].q

    def known_removed(self, viewer: int) -> set[str]:
        """Creature ids the viewer knows are out of play."""
        removed = set()
        for ev in self.log:
            if ev[0] == "identified":
                removed.add(ev[3])
        removed.add(self.players[viewer].specimen.cid)  # my own can't be anyone else's
        return removed

    def candidates(self, viewer: int, rival: int) -> list[Creature]:
        """All creatures consistent with everything `viewer` knows about `rival`."""
        p = self.players[rival]
        removed = self.known_removed(viewer)
        cands = [c for c in self.creatures if c.cid not in removed]
        for ev in self.log:
            kind = ev[0]
            if kind in ("answer", "reveal", "guess_fail") and ev[1] == rival and ev[2] == p.epoch:
                if kind == "answer":
                    _, _t, _e, trait, value, yes, private_to = ev
                    if private_to is not None and private_to != viewer:
                        continue
                    cands = [c for c in cands if c.has(trait, value) == yes]
                elif kind == "reveal":
                    _, _t, _e, trait, value, private_to = ev
                    if private_to is not None and private_to != viewer:
                        continue
                    cands = [c for c in cands if c.value_of(trait) == value]
                else:  # guess_fail: public information
                    cid = ev[3]
                    cands = [c for c in cands if c.cid != cid]
        return cands

    # ------------------------------------------------------------------ turns

    def _draw(self, p: PlayerState):
        if not self.question_deck and self.question_discard:
            self.question_deck = self.question_discard
            self.question_discard = []
            self.rng.shuffle(self.question_deck)
        if self.question_deck:
            p.hand.append(self.question_deck.pop())
        while len(p.hand) > HAND_LIMIT:  # enforce hand limit deterministically
            self.question_discard.append(p.hand.pop(0))

    def _resolve(self, p: PlayerState, act: tuple):
        kind = act[0]
        if kind == "ask_card":
            _, card, target = act
            if card not in p.hand or target == p.seat or card.special:
                return
            p.hand.remove(card)
            self.question_discard.append(card)
            self._answer(p.seat, target, card.trait, card.value, private_to=None)
        elif kind == "ask_free":
            _, trait, value, target, private_to = act
            if target == p.seat or trait not in TRAITS:
                return
            self._answer(p.seat, target, trait, value, private_to=private_to)
        elif kind == "reveal_trait":
            _, trait, target, private_to = act
            if target == p.seat or trait not in TRAITS:
                return
            t = self.players[target]
            self.log.append(("reveal", target, t.epoch, trait, t.specimen.value_of(trait), private_to))
        elif kind == "play_special":
            _, card = act
            if card in p.hand and card.special:
                p.hand.remove(card)
                self.question_discard.append(card)
        elif kind == "guess":
            _, target, cid = act
            if target == p.seat:
                return
            self.guess_attempts += 1
            t = self.players[target]
            if t.specimen.cid == cid:
                self.guess_hits += 1
                self.q_at_hit.append(t.q)
                points = 2 + max(0, 4 - t.q)
                p.score += points
                p.correct_ids += 1
                self.log.append(("identified", target, t.epoch, cid, t.q))
                if self.creature_deck:
                    t.specimen = self.creature_deck.pop()
                    t.epoch += 1
                    t.q = 0
                else:
                    self.depleted = True
            else:
                p.score = max(0, p.score - 2)
                self.log.append(("guess_fail", target, t.epoch, cid))
        elif kind == "discard":
            _, card = act
            if card in p.hand:
                p.hand.remove(card)
                self.question_discard.append(card)

    def _answer(self, asker: int, target: int, trait: str, value: str, private_to: int | None):
        t = self.players[target]
        yes = t.specimen.has(trait, value)
        self.log.append(("answer", target, t.epoch, trait, value, yes, private_to))
        if private_to is None:
            t.q += 1

    def play(self, on_turn=None) -> GameResult:
        while not self.over and self.turns < MAX_TURNS:
            self.turn_seat = self.order[self.pos]
            p = self.players[self.turn_seat]
            self._draw(p)
            log_before = len(self.log)
            actions = self.agents[self.turn_seat].play(self, self.turn_seat)
            resolved_main = False
            for act in actions:
                if act[0] in ("ask_card", "guess", "discard") and resolved_main:
                    continue
                self._resolve(p, act)
                if act[0] in ("ask_card", "guess", "discard", "play_special"):
                    resolved_main = True
            if not resolved_main and p.hand:  # forced discard if agent idled
                self.question_discard.append(p.hand.pop(0))
            if on_turn:
                on_turn(self, p.seat, actions, self.log[log_before:])

            if p.score >= self.win_target:
                self.finish_triggered = True  # equal-turn finish: complete the round

            self.turns += 1
            self.pos += 1
            if self.pos == self.n:  # round complete
                self.pos = 0
                if self.rotate:
                    # pass the first-player marker: cancels the late-seat
                    # listening advantage over the course of the game.
                    self.order = self.order[1:] + self.order[:1]
                if self.depleted or self.finish_triggered:
                    self.over = True
                    self.winner = self._settle_by_score()
            self.turn_seat = self.order[self.pos]

        if not self.over:
            self.capped = True
            self.over = True
            self.winner = self._settle_by_score()

        return GameResult(
            winner=self.winner,
            turns=self.turns,
            scores=[pl.score for pl in self.players],
            guess_attempts=self.guess_attempts,
            guess_hits=self.guess_hits,
            q_at_hit=self.q_at_hit,
            capped=self.capped,
        )

    def _settle_by_score(self) -> int:
        best = max((pl.score, pl.correct_ids, -pl.seat) for pl in self.players)
        for pl in self.players:
            if (pl.score, pl.correct_ids, -pl.seat) == best:
                return pl.seat
        return 0

File: games/spotted/web/src/game/engine.ts
/**
 * SPOTTED! rules engine — faithful TypeScript port of sim/engine.py.
 * Implements RULES.md exactly:
 * - draw 1, then one action (question / special / identification / discard)
 * - public answers increment the target's q counter (efficiency penalty meter)
 * - scoring: 2 + max(0, 4 - q) for a correct identification
 * - win at winTarget tokens; equal-turn finish; tiebreaks; first-player rotation
 */
import {
  Creature, QuestionCard, Trait, TRAITS,
  CREATURES, buildQuestionDeck, hasTrait,
} from "./data";
import { makeRng, Rng } from "./rng";

export const HAND_LIMIT = 7;
export const STARTING_HAND = 6;
export const WIN_TARGET_DEFAULT = 10;
export const WIN_TARGET_2P = 8; // Duel variant
export const WIN_TARGET_KIDS = 6;
export const MAX_TURNS = 400;

export interface PlayerState {
  seat: number;
  hand: QuestionCard[];
  specimen: Creature;
  epoch: number; // increments every time this player's specimen is replaced
  score: number;
  q: number; // public answers about current specimen
  correctIds: number;
}

export type LogEvent =
  | { kind: "answer"; target: number; epoch: number; trait: Trait; value: string; yes: boolean; privateTo: number | null }
  | { kind: "reveal"; target: number; epoch: number; trait: Trait; value: string; privateTo: number }
  | { kind: "guess_fail"; target: number; epoch: number; cid: string }
  | { kind: "identified"; target: number; epoch: number; cid: string; q: number };

export type Action =
  | { kind: "ask_card"; card: QuestionCard; target: number }
  | { kind: "ask_free"; trait: Trait; value: string; target: number; privateTo: number | null }
  | { kind: "reveal_trait"; trait: Trait; target: number; privateTo: number }
  | { kind: "play_special"; card: QuestionCard }
  | { kind: "guess"; target: number; cid: string }
  | { kind: "discard"; card: QuestionCard };

export interface GameResult {
  winner: number; // seat, -1 for capped/abnormal
  turns: number;
  scores: number[];
  guessAttempts: number;
  guessHits: number;
  capped: boolean;
}

export interface GameConfig {
  nPlayers: number;
  seed: number;
  kidsMode?: boolean;
  winTarget?: number;
  rotateFirstPlayer?: boolean;
}

export interface Agent {
  play(game: Game, seat: number): Action[];
}

export class Game {
  rng: Rng;
  n: number;
  agents: Agent[] = [];
  rotate: boolean;
  kidsMode: boolean;
  winTarget: number;
  creatures: Creature[] = CREATURES;
  creatureDeck: Creature[];
  questionDeck: QuestionCard[];
  questionDiscard: QuestionCard[] = [];
  players: PlayerState[];
  log: LogEvent[] = [];
  depleted = false;
  finishTriggered = false;
  turns = 0;
  order: number[];
  pos = 0;
  turnSeat = 0;
  guessAttempts = 0;
  guessHits = 0;
  over = false;
  winner = -1;
  capped = false;
  /** true between beginTurn() and endTurn() — a turn is in progress */
  inTurn = false;
  lastDrawn: QuestionCard | null = null;

  constructor(cfg: GameConfig) {
    const n = cfg.nPlayers;
    if (n < 2 || n > 5) throw new Error("2-5 players");
    this.rng = makeRng(cfg.seed);
    this.n = n;
    this.rotate = cfg.rotateFirstPlayer ?? true;
    this.kidsMode = cfg.kidsMode ?? false;
    this.winTarget = cfg.winTarget ?? (this.kidsMode ? WIN_TARGET_KIDS
      : n === 2 ? WIN_TARGET_2P : WIN_TARGET_DEFAULT);
    this.creatureDeck = this.rng.shuffle([...this.creatures]);
    this.questionDeck = this.rng.shuffle(buildQuestionDeck());

    this.players = [];
    for (let i = 0; i < n; i++) {
      const hand: QuestionCard[] = [];
      for (let k = 0; k < STARTING_HAND; k++) hand.push(this.questionDeck.pop()!);
      this.players.push({
        seat: i, hand, specimen: this.creatureDeck.pop()!,
        epoch: 0, score: 0, q: 0, correctIds: 0,
      });
    }
    this.order = [...Array(n).keys()];
  }

  // ------------------------------------------------------------------ facts

  publicQ(seat: number): number {
    return this.players[seat].q;
  }

  /** Creature ids the viewer knows are out of play. */
  knownRemoved(viewer: number): Set<string> {
    const removed = new Set<string>();
    for (const ev of this.log) if (ev.kind === "identified") removed.add(ev.cid);
    removed.add(this.players[viewer].specimen.cid); // my own can't be anyone else's
    return removed;
  }

  /** All creatures consistent with everything `viewer` knows about `rival`. */
  candidates(viewer: number, rival: number): Creature[] {
    const p = this.players[rival];
    const removed = this.knownRemoved(viewer);
    let cands = this.creatures.filter((c) => !removed.has(c.cid));
    for (const ev of this.log) {
      if (ev.target !== rival || ev.epoch !== p.epoch) continue;
      if (ev.kind === "answer") {
        if (ev.privateTo !== null && ev.privateTo !== viewer) continue;
        cands = cands.filter((c) => hasTrait(c, ev.trait, ev.value) === ev.yes);
      } else if (ev.kind === "reveal") {
        if (ev.privateTo !== viewer) continue;
        cands = cands.filter((c) => c.traits[ev.trait] === ev.value);
      } else if (ev.kind === "guess_fail") {
        cands = cands.filter((c) => c.cid !== ev.cid);
      }
    }
    return cands;
  }

  // ------------------------------------------------------------------ turns

  /** Step 1 — draw (reshuffle discard if empty; enforce hand limit). */
  beginTurn(): QuestionCard | null {
    if (this.over || this.inTurn) throw new Error("bad turn state");
    this.inTurn = true;
    this.turnSeat = this.order[this.pos];
    const p = this.players[this.turnSeat];
    this.lastDrawn = null;
    if (this.questionDeck.length === 0 && this.questionDiscard.length) {
      this.questionDeck = this.rng.shuffle(this.questionDiscard);
      this.questionDiscard = [];
    }
    if (this.questionDeck.length) {
      this.lastDrawn = this.questionDeck.pop()!;
      p.hand.push(this.lastDrawn);
    }
    while (p.hand.length > HAND_LIMIT) this.questionDiscard.push(p.hand.shift()!);
    return this.lastDrawn;
  }

  /** Step 2 — resolve the chosen actions (exactly one main action enforced by UI/agents). */
  resolve(actions: Action[]): LogEvent[] {
    const before = this.log.length;
    const p = this.players[this.turnSeat];
    let resolvedMain = false;
    for (const act of actions) {
      if (resolvedMain && (act.kind === "ask_card" || act.kind === "guess" || act.kind === "discard")) continue;
      this.resolveOne(p, act);
      if (act.kind === "ask_card" || act.kind === "guess" || act.kind === "discard" || act.kind === "play_special") {
        resolvedMain = true;
      }
    }
    if (!resolvedMain && p.hand.length) this.questionDiscard.push(p.hand.shift()!);
    return this.log.slice(before);
  }

  private resolveOne(p: PlayerState, act: Action): void {
    switch (act.kind) {
      case "ask_card": {
        const { card, target } = act;
        if (!p.hand.includes(card) || target === p.seat || card.special) return;
        p.hand.splice(p.hand.indexOf(card), 1);
        this.questionDiscard.push(card);
        this.answer(p.seat, target, card.trait as Trait, card.value, null);
        break;
      }
      case "ask_free": {
        const { trait, value, target, privateTo } = act;
        if (target === p.seat || !(TRAITS as readonly string[]).includes(trait)) return;
        this.answer(p.seat, target, trait, value, privateTo);
        break;
      }
      case "reveal_trait": {
        const { trait, target, privateTo } = act;
        if (target === p.seat || !(TRAITS as readonly string[]).includes(trait)) return;
        const t = this.players[target];
        this.log.push({ kind: "reveal", target, epoch: t.epoch, trait, value: t.specimen.traits[trait], privateTo });
        break;
      }
      case "play_special": {
        const { card } = act;
        if (p.hand.includes(card) && card.special) {
          p.hand.splice(p.hand.indexOf(card), 1);
          this.questionDiscard.push(card);
        }
        break;
      }
      case "guess": {
        const { target, cid } = act;
        if (target === p.seat) return;
        this.guessAttempts += 1;
        const t = this.players[target];
        if (t.specimen.cid === cid) {
          this.guessHits += 1;
          const points = this.kidsMode ? 2 : 2 + Math.max(0, 4 - t.q);
          p.score += points;
          p.correctIds += 1;
          this.log.push({ kind: "identified", target, epoch: t.epoch, cid, q: t.q });
          if (this.creatureDeck.length) {
            t.specimen = this.creatureDeck.pop()!;
            t.epoch += 1;
            t.q = 0;
          } else {
            this.depleted = true;
          }
        } else {
          if (!this.kidsMode) p.score = Math.max(0, p.score - 3);
          this.log.push({ kind: "guess_fail", target, epoch: t.epoch, cid });
        }
        break;
      }
      case "discard": {
        const { card } = act;
        if (p.hand.includes(card)) {
          p.hand.splice(p.hand.indexOf(card), 1);
          this.questionDiscard.push(card);
        }
        break;
      }
    }
  }

  private answer(asker: number, target: number, trait: Trait, value: string, privateTo: number | null): void {
    const t = this.players[target];
    const yes = hasTrait(t.specimen, trait, value);
    this.log.push({ kind: "answer", target, epoch: t.epoch, trait, value, yes, privateTo });
    if (privateTo === null) t.q += 1;
  }

  /** End of turn: win trigger check, rotation, equal-turn finish. */
  endTurn(): void {
    if (!this.inTurn) throw new Error("endTurn without beginTurn");
    this.inTurn = false;
    const p = this.players[this.turnSeat];
    if (p.score >= this.winTarget) this.finishTriggered = true;

    this.turns += 1;
    this.pos += 1;
    if (this.pos === this.n) {
      this.pos = 0;
      if (this.rotate) {
        // pass the first-player marker: cancels the late-seat listening advantage
        this.order = [...this.order.slice(1), this.order[0]];
      }
      if (this.depleted || this.finishTriggered) {
        this.over = true;
        this.winner = this.settleByScore();
      }
    }
    if (!this.over && this.turns >= MAX_TURNS) {
      this.capped = true;
      this.over = true;
      this.winner = this.settleByScore();
    }
  }

  private settleByScore(): number {
    let bestSeat = 0;
    let best: [number, number, number] = [-1, -1, 0];
    for (const pl of this.players) {
      const key: [number, number, number] = [pl.score, pl.correctIds, -pl.seat];
      if (key[0] > best[0] || (key[0] === best[0] && key[1] > best[1]) ||
          (key[0] === best[0] && key[1] === best[1] && key[2] > best[2])) {
        best = key; bestSeat = pl.seat;
      }
    }
    return bestSeat;
  }

  result(): GameResult {
    return {
      winner: this.winner,
      turns: this.turns,
      scores: this.players.map((p) => p.score),
      guessAttempts: this.guessAttempts,
      guessHits: this.guessHits,
      capped: this.capped,
    };
  }

  /** Headless full game (used by tests + bot-vs-bot smoke runs). */
  play(onTurn?: (game: Game, seat: number, actions: Action[], events: LogEvent[]) => void): GameResult {
    while (!this.over) {
      this.beginTurn();
      const seat = this.turnSeat;
      const actions = this.agents[seat].play(this, seat);
      const events = this.resolve(actions);
      if (onTurn) onTurn(this, seat, actions, events);
      this.endTurn();
    }
    return this.result();
  }
}
